File: 61-rotate-list/rotate-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if(head==nullptr || head->next==nullptr || k==0)
        {
            return head;
        }
        int length=0;
        ListNode* curr=head;
        while(curr!=nullptr)
        {
            curr=curr->next;
            length+=1;
        }
        k=k%length;
        if(k==0)
        {
            return head;
        }
        ListNode* slow=head;
        ListNode* fast=head;
        for(int i=0;i<k;i++)
        {
            fast=fast->next;
        }
        while(fast->next!=nullptr)
        {
            slow=slow->next;
            fast=fast->next;
        }
        ListNode* newhead=slow->next;
        slow->next=nullptr;
        fast->next=head;
        return newhead;

    }
};

File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head or not head.next or k==0:
            return head
        length=0
        curr=head
        while(curr!=None):
            curr=curr.next
            length+=1
        k=k%length
        if(k==0):
            return head
        s=head
        f=head
        for i in range(k):
            f=f.next
        while(f.next!=None):
            s=s.next
            f=f.next
        newhead=s.next
        s.next=None
        f.next=head
        return newhead

        